From a52b22c4010279acaa6cf0dc7f4808826e7cc8e4 Mon Sep 17 00:00:00 2001
From: Luca Di Maio <luca.dimaio1@gmail.com>
Date: Fri, 5 Dec 2025 19:03:17 +0100
Subject: [PATCH v0] fix: add policies for overlay dirs for immutable builds

we need proper policies for these mounts, the mapping is not enough

Ref VMS-81
Ref VMS-161

Signed-off-by: Luca Di Maio <luca.dimaio1@gmail.com>
---
 policy/modules/kernel/files.fc | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/policy/modules/kernel/files.fc b/policy/modules/kernel/files.fc
index e898cb31f..85009b54f 100644
--- a/policy/modules/kernel/files.fc
+++ b/policy/modules/kernel/files.fc
@@ -338,3 +338,26 @@ ifndef(`distro_redhat',`
 
 /sysroot/.aleph-version.json	gen_context(system_u:object_r:root_t,s0)
 /sysroot/ostree/deploy/.*-atomic/deploy(/.*)?           gen_context(system_u:object_r:root_t,s0)
+
+# Overlay upperdirs for immutable/volatile systems
+/run/overlays/etc/upper			-d	gen_context(system_u:object_r:etc_t,s0)
+/run/overlays/etc/upper(/.*)?			gen_context(system_u:object_r:etc_t,s0)
+/run/overlays/etc/work			-d	gen_context(system_u:object_r:etc_t,s0)
+/run/overlays/var/upper			-d	gen_context(system_u:object_r:var_t,s0)
+/run/overlays/var/upper(/.*)?			gen_context(system_u:object_r:var_t,s0)
+/run/overlays/var/work			-d	gen_context(system_u:object_r:var_t,s0)
+# /home needs to be home_root_t while in the upper we have user dirs, so we need them
+# to be user_home_dir_t but files inside user_home_dir_t need to inherit label from policy
+# mechanism (eg: ssh_home_t, gpg_secret_t)
+/run/overlays/home/upper		-d	gen_context(system_u:object_r:home_root_t,s0)
+/run/overlays/home/upper/[^/]+		-d	gen_context(system_u:object_r:user_home_dir_t,s0)
+/run/overlays/home/upper/[^/]+/.*		<<none>>
+/run/overlays/home/work			-d	gen_context(system_u:object_r:home_root_t,s0)
+# likewise for root
+/run/overlays/root/upper		-d	gen_context(system_u:object_r:admin_home_t,s0)
+/run/overlays/root/upper(/.*)?			<<none>>
+/run/overlays/root/work			-d	gen_context(system_u:object_r:admin_home_t,s0)
+# likewise for usr/local
+/run/overlays/usr/local/upper		-d	gen_context(system_u:object_r:usr_t,s0)
+/run/overlays/usr/local/upper(/.*)?		<<none>>
+/run/overlays/usr/local/work		-d	gen_context(system_u:object_r:usr_t,s0)
-- 
2.52.0

